src/modules/ekf2/EKF/imu_down_sampler.cpp

```diff
@@ -33,33 +36,31 @@ bool ImuDownSampler::update(const imuSample &imu_sample_new)
 	// assume effective sample time is halfway between the previous and current rotation frame
 	_imu_down_sampled.delta_vel += (imu_sample_new.delta_vel + delta_R * imu_sample_new.delta_vel) * 0.5f;
 
-	// check if the target time delta between filter prediction steps has been exceeded
-	if (_imu_down_sampled.delta_ang_dt >= _target_dt - _imu_collection_time_adj) {
-		// accumulate the amount of time to advance the IMU collection time so that we meet the
-		// average EKF update rate requirement
-		_imu_collection_time_adj += 0.01f * (_imu_down_sampled.delta_ang_dt - _target_dt);
-		_imu_collection_time_adj = math::constrain(_imu_collection_time_adj, -0.5f * _target_dt,
-					   0.5f * _target_dt);
+	_accumulated_samples++;
 
-		_imu_down_sampled.delta_ang = AxisAnglef(_delta_angle_accumulated);
 
-		return true;
+	// required number of samples accumulated and the total time is at least half of the target
+	//  OR total time already exceeds the target
+	if ((_accumulated_samples >= _required_samples && _imu_down_sampled.delta_ang_dt > 0.5f * _target_dt_s)
```

Comment:
What error is handled by including an accumulated ample condition rather than the delta angle? is this for compatibility with downstream consumers?

Reply:
That's a good question I'm not sure I have a proper answer for. I found it easier to keep things on track running synchronized with IMU, but part of that might be the way I'm thinking about the problem. The sample count is the expected case, the accumulated delta angle dt checks are then protections and sanity checks.

Because most of our sensors don't share clocks with the flight controller our IMU data is almost never perfectly aligned with the 10 millisecond filter update period. Even with very stable IMU delta angle dt input the existing IMU down sampler logic would accumulate enough collection time adjustment to periodically skip a sample, introducing real jitter in the down sampled IMU and scheduling.

![Screenshot from 2022-01-20 10-04-01](https://user-images.githubusercontent.com/84712/150364629-31a61600-479c-45b2-abf4-d0b3b9e7ed6b.png)

Reply:
That makes sense. Keeping a consistent downsample ratio where possible is preferred for minimising  timing jitter, but will cause an occasional reduced EKF internal time step which the EKF can tolerate. However we need to be aware that if the EKF's average time step is reduced that does reduce stability margins for some covariance processing operations that have dt^2 in them.

Reply:
If it's a concern we could further tighten the secondary condition (_imu_down_sampled.delta_ang_dt > 0.5f * _target_dt_s). We can also be more opinionated when selecting the actual IMU target rate, maybe rounding up for the cases that don't align well with 10 milliseconds. For example the 250 Hz IMU data in SITL results in an EKF dt of 8 ms, but it could instead be 12 ms.

https://github.com/PX4/PX4-Autopilot/blob/a95b1156156d68a340cdb156b333f37a31a635f7/src/modules/ekf2/EKF/imu_down_sampler.cpp#L63

My goal is to minimize jitter across the system, and more fundamentally align with how the hardware actually operates. The "good" cases (eg FIFO IMU with watermark interrupt) are nearly jitter free now, but the clocks are slightly different (< 0.5%). There are other cases where the IMU scheduling is more erratic, and the integration layer (sensors/vehicle_imu) does it's best to provide clean output, but it's still jumping around a bit.